File: OhHckGame.java
import java.util.ArrayList;
import java.util.List;

public class OhHckGame {

    public enum State {
        WAITING, STARTED, BIDDING, TRICK
    }

    private State state;
    private List<OhHckServer.ServerThread> clients;
    private int upperLimit;
    private Deck deck;
    private int dealer;
    private int numCards;
    private int bidTotal;
    private int plon;
    private List<TrackingCard> played;
    private int ctr;
    private boolean direction;
    private int nsp;

    public OhHckGame() {
        this.state = State.WAITING;
        this.clients = new ArrayList<OhHckServer.ServerThread>();
        this.numCards = 1;
        this.played = new ArrayList<TrackingCard>();
        this.direction = true;
    }

    public void addClient(OhHckServer.ServerThread client) {
        clients.add(client);
    }

    // SCRAPPY STATE MACHINE
    public void process(String input, OhHckServer.ServerThread sender) {
        if (input == null) {
            if (state == State.WAITING) {
                sender.player().setPlayerName(
                    "Player" + sender.getCurrentPlayers());
                sender.transmit("WELCOME " + sender.player().getPlayerName());
            } else {
                sender.transmit("IN PROGRESS");
            }
        } else if (input.equals("STOP") && sender == clients.get(0)) {
            transmitAll("STOP");
        } else {
            try {
                boolean matched = false;
                switch (state) {
                    case WAITING:
                        if (input.length() >= 6
                                && input.substring(0, 6).equals("START ")
                                && sender == clients.get(0) && clients.size() >= 3) {
                            upperLimit = Integer.parseInt(
                                input.substring(input.lastIndexOf(' ') + 1));
                            deck = new Deck(Integer.parseInt(input.substring(
                                input.indexOf(' ') + 1, input.indexOf(' ') + 2)));
                            matched = true;
                            state = State.STARTED;
                            transmitAll(scores());
                            dealer = (int) (Math.random() * clients.size());
                            clients.get(dealer).transmit("DEALER");
                        }
                        break;
                    case STARTED:
                        if (input.equals("BEGDEAL") && sender == clients.get(dealer)) {
                            matched = true;
                            state = State.BIDDING;
                            for (int i = 0; i < numCards; i++) {
                                for (int j = 0; j < clients.size(); j++) {
                                    clients.get((dealer + j + 1) % clients.size())
                                        .transmit("DEAL " + deck.pop().toString());
                                }
                            }
                            eraseBids();
                            transmitAll(bids());
                            clients.get((dealer + 1) % clients.size())
                                .transmit("PLACEBID " + bidRestriction(dealer + 1));
                            plon = dealer + 1;
                            nsp = plon % clients.size();
                        }
                        break;
                    case BIDDING:
                        if (input.length() >= 4
                                && input.substring(0, 4).equals("BID ")
                                && sender == clients.get(plon % clients.size())) {
                            sender.player().setBid(Integer.parseInt(input.substring(4)));
                            updateBidTotal();
                            plon++;
                            matched = true;
                            transmitAll(bids());
                            if (plon < dealer + clients.size() + 1) {
                                clients.get(plon % clients.size())
                                    .transmit("PLACEBID " + bidRestriction(plon));
                            } else {
                                state = State.TRICK;
                                ctr = 0;
                                clearTricks();
                                played.clear();
                                transmitAll("PLAYED " + played);
                                clients.get(plon % clients.size()).transmit("PLAY");
                            }
                        }
                        break;
                    case TRICK:
                        if (input.length() >= 8
                                && input.substring(0, 8).equals("PLAYING ")
                                && sender == clients.get(plon % clients.size())) {
                            played.add(new TrackingCard(
                                Card.fromString(input.substring(8)), sender,
                                System.currentTimeMillis()));
                            plon++;
                            matched = true;
                            transmitAll("PLAYED " + played);
                            if (plon % clients.size() != nsp) {
                                clients.get(plon % clients.size()).transmit("PLAY");
                            } else {
                                transmitAll("TRICK " + winner());
                                ctr++;
                                if (ctr < numCards) {
                                    played.clear();
                                    transmitAll("PLAYED " + played);
                                    clients.get(plon % clients.size()).transmit("PLAY");
                                } else {
                                    ctr = 0;
                                    if (numCards == upperLimit) {
                                        direction = false;
                                    }
                                    numCards += (direction) ? 1 : -1;
                                    assignScores();
                                    transmitAll(scores());
                                    if (numCards != 0) {
                                        state = State.STARTED;
                                        deck.reset();
                                        dealer++;
                                        dealer %= clients.size();
                                        clients.get(dealer).transmit("DEALER");
                                    } else {
                                        transmitAll(results());
                                        transmitAll("STOP");
                                    }
                                }
                            }
                        }
                        break;
                }
                if (!matched) {
                    sender.transmit("NOT RECOGNIZED");
                }
            } catch (Throwable t) {
                sender.transmit("ERROR " + t.getClass().getName());
            }
        }
    }

    private void assignScores() {
        for (OhHckServer.ServerThread c : clients) {
            if (c.player().getTricks() == c.player().getBid()) {
                c.player().addScore(10 + c.player().getBid());
            }
        }
    }

    private String winner() {
        Card.Suit led = played.get(0).card.getSuit();
        played.sort((c1, c2) -> {
            Card.Suit s1 = c1.card.getSuit();
            Card.Suit s2 = c2.card.getSuit();
            if (s1 == Card.Suit.SPADES && s2 != Card.Suit.SPADES) {
                return 69;
            } else if (s1 != Card.Suit.SPADES && s2 == Card.Suit.SPADES) {
                return -69;
            } else if (s1 == led && s2 != led) {
                return 24;
            } else if (s1 != led && s2 == led) {
                return -24;
            } else {
                int diff = c1.card.getRank() - c2.card.getRank();
                return (diff == 0)
                    ? (int) ((c2.timePlayed - c1.timePlayed) / 1000) : diff;
            }
        });
        OhHckServer.ServerThread t = played.get(played.size() - 1).playedBy;
        t.player().setTricks(true);
        plon = clients.indexOf(t);
        nsp = plon;
        return t.player().getPlayerName();
    }

    private int bidRestriction(int clindex) {
        return (clindex % clients.size() == dealer) ? numCards - bidTotal : -1;
    }

    private String results() {
        return scores(true);
    }

    private String scores() {
        return scores(false);
    }

    private String scores(boolean isFinal) {
        StringBuilder ret = new StringBuilder();
        if (!isFinal) {
            ret.append("SCORES [");
        } else {
            ret.append("RESULTS [");
        }
        for (OhHckServer.ServerThread c : clients) {
            ret.append(c.player().getPlayerName());
            ret.append('=');
            ret.append(c.player().getScore());
            ret.append(", ");
        }
        return ret.substring(0, ret.length() - 2) + ']';
    }

    private void updateBidTotal() {
        bidTotal = 0;
        for (OhHckServer.ServerThread c : clients) {
            if (c.player().getBid() != -1) {
                bidTotal += c.player().getBid();
            }
        }
    }

    private void eraseBids() {
        for (OhHckServer.ServerThread c : clients) {
            c.player().setBid(-1);
        }
        updateBidTotal();
    }

    private void clearTricks() {
        for (OhHckServer.ServerThread c : clients) {
            c.player().setTricks(false);
        }
    }

    private String bids() {
        StringBuilder ret = new StringBuilder();
        ret.append("BIDS [");
        for (OhHckServer.ServerThread c : clients) {
            ret.append(c.player().getPlayerName());
            ret.append('=');
            ret.append(c.player().getBid());
            ret.append(", ");
        }
        return ret.substring(0, ret.length() - 2) + ']';
    }

    private void transmitAll(String message) {
        for (OhHckServer.ServerThread c : clients) {
            c.transmit(message);
        }
    }

    public class TrackingCard {

        private Card card;
        private OhHckServer.ServerThread playedBy;
        private long timePlayed;

        public TrackingCard(Card card, OhHckServer.ServerThread playedBy,
                long timePlayed) {
            this.card = card;
            this.playedBy = playedBy;
            this.timePlayed = timePlayed;
        }

        @Override
        public String toString() {
            return playedBy.player().getPlayerName() + "=" + card.toString();
        }
    }

}

File: src/main/java/com/isaac/games/OhHckGame.kt
package com.isaac.games

class OhHckGame {

    private var state: State? = null
    private val clients: MutableList<OhHckServer.ServerThread>
    private var upperLimit: Int = 0
    private var deck: Deck? = null
    private var dealer: Int = 0
    private var numCards: Int = 0
    private var bidTotal: Int = 0
    private var plon: Int = 0
    private val played: MutableList<TrackingCard>
    private var ctr: Int = 0
    private var direction: Boolean = false
    private var nsp: Int = 0

    enum class State {
        WAITING, STARTED, BIDDING, TRICK
    }

    init {
        this.state = State.WAITING
        this.clients = ArrayList()
        this.numCards = 1
        this.played = ArrayList()
        this.direction = true
    }

    fun addClient(client: OhHckServer.ServerThread) {
        clients.add(client)
    }

    // SCRAPPY STATE MACHINE
    fun process(input: String?, sender: OhHckServer.ServerThread) {
        if (input == null) {
            if (state == State.WAITING) {
                sender.player().playerName = "Player" + sender.currentPlayers
                sender.transmit("WELCOME " + sender.player().playerName)
            } else {
                sender.transmit("IN PROGRESS")
            }
        } else if (input == "STOP" && sender === clients[0]) {
            transmitAll("STOP")
        } else {
            try {
                var matched = false
                when (state) {
                    OhHckGame.State.WAITING -> if (input.length >= 6
                            && input.substring(0, 6) == "START "
                            && sender === clients[0] && clients.size >= 3) {
                        upperLimit = Integer.parseInt(
                                input.substring(input.lastIndexOf(' ') + 1))
                        deck = Deck(Integer.parseInt(input.substring(
                                input.indexOf(' ') + 1, input.indexOf(' ') + 2)))
                        matched = true
                        state = State.STARTED
                        transmitAll(scores())
                        dealer = (Math.random() * clients.size).toInt()
                        clients[dealer].transmit("DEALER")
                    }
                    OhHckGame.State.STARTED -> if (input == "BEGDEAL" && sender === clients[dealer]) {
                        matched = true
                        state = State.BIDDING
                        for (i in 0 until numCards) {
                            for (j in clients.indices) {
                                clients[(dealer + j + 1) % clients.size]
                                        .transmit("DEAL " + deck!!.pop().toString())
                            }
                        }
                        eraseBids()
                        transmitAll(bids())
                        clients[(dealer + 1) % clients.size]
                                .transmit("PLACEBID " + bidRestriction(dealer + 1))
                        plon = dealer + 1
                        nsp = plon % clients.size
                    }
                    OhHckGame.State.BIDDING -> if (input.length >= 4
                            && input.substring(0, 4) == "BID "
                            && sender === clients[plon % clients.size]) {
                        sender.player().bid = Integer.parseInt(input.substring(4))
                        updateBidTotal()
                        plon++
                        matched = true
                        transmitAll(bids())
                        if (plon < dealer + clients.size + 1) {
                            clients[plon % clients.size]
                                    .transmit("PLACEBID " + bidRestriction(plon))
                        } else {
                            state = State.TRICK
                            ctr = 0
                            clearTricks()
                            played.clear()
                            transmitAll("PLAYED $played")
                            clients[plon % clients.size].transmit("PLAY")
                        }
                    }
                    OhHckGame.State.TRICK -> if (input.length >= 8
                            && input.substring(0, 8) == "PLAYING "
                            && sender === clients[plon % clients.size]) {
                        played.add(TrackingCard(
                                Card.fromString(input.substring(8)), sender,
                                System.currentTimeMillis()))
                        plon++
                        matched = true
                        transmitAll("PLAYED $played")
                        if (plon % clients.size != nsp) {
                            clients[plon % clients.size].transmit("PLAY")
                        } else {
                            transmitAll("TRICK " + winner())
                            ctr++
                            if (ctr < numCards) {
                                played.clear()
                                transmitAll("PLAYED $played")
                                clients[plon % clients.size].transmit("PLAY")
                            } else {
                                ctr = 0
                                if (numCards == upperLimit) {
                                    direction = false
                                }
                                numCards += if (direction) 1 else -1
                                assignScores()
                                transmitAll(scores())
                                if (numCards != 0) {
                                    state = State.STARTED
                                    deck!!.reset()
                                    dealer++
                                    dealer %= clients.size
                                    clients[dealer].transmit("DEALER")
                                } else {
                                    transmitAll(results())
                                    transmitAll("STOP")
                                }
                            }
                        }
                    }
                }
                if (!matched) {
                    sender.transmit("NOT RECOGNIZED")
                }
            } catch (t: Throwable) {
                sender.transmit("ERROR " + t.javaClass.name)
            }

        }
    }

    private fun assignScores() {
        for (c in clients) {
            if (c.player().tricks == c.player().bid) {
                c.player().addScore(10 + c.player().bid)
            }
        }
    }

    private fun winner(): String {
        val led = played[0].card.suit
        played.sortWith(Comparator { c1, c2 ->
            val s1 = c1.card.suit
            val s2 = c2.card.suit
            if (s1 === Card.Suit.SPADES && s2 !== Card.Suit.SPADES) {
                69
            } else if (s1 !== Card.Suit.SPADES && s2 === Card.Suit.SPADES) {
                -69
            } else if (s1 === led && s2 !== led) {
                24
            } else if (s1 !== led && s2 === led) {
                -24
            } else {
                val diff = c1.card.rank - c2.card.rank
                if (diff == 0)
                    ((c2.timePlayed - c1.timePlayed) / 1000).toInt()
                else
                    diff
            }
        })
        val t = played[played.size - 1].playedBy
        t.player().setTricks(true)
        plon = clients.indexOf(t)
        nsp = plon
        return t.player().playerName!!
    }

    private fun bidRestriction(clindex: Int): Int {
        return if (clindex % clients.size == dealer) numCards - bidTotal else -1
    }

    private fun results(): String {
        return scores(true)
    }

    private fun scores(isFinal: Boolean = false): String {
        val ret = StringBuilder()
        if (!isFinal) {
            ret.append("SCORES [")
        } else {
            ret.append("RESULTS [")
        }
        for (c in clients) {
            ret.append(c.player().playerName)
            ret.append('=')
            ret.append(c.player().score)
            ret.append(", ")
        }
        return ret.substring(0, ret.length - 2) + ']'
    }

    private fun updateBidTotal() {
        bidTotal = 0
        for (c in clients) {
            if (c.player().bid != -1) {
                bidTotal += c.player().bid
            }
        }
    }

    private fun eraseBids() {
        for (c in clients) {
            c.player().bid = -1
        }
        updateBidTotal()
    }

    private fun clearTricks() {
        for (c in clients) {
            c.player().setTricks(false)
        }
    }

    private fun bids(): String {
        val ret = StringBuilder()
        ret.append("BIDS [")
        for (c in clients) {
            ret.append(c.player().playerName)
            ret.append('=')
            ret.append(c.player().bid)
            ret.append(", ")
        }
        return ret.substring(0, ret.length - 2) + ']'
    }

    private fun transmitAll(message: String) {
        for (c in clients) {
            c.transmit(message)
        }
    }

    inner class TrackingCard(val card: Card, val playedBy: OhHckServer.ServerThread,
                             val timePlayed: Long) {

        override fun toString(): String {
            return playedBy.player().playerName + "=" + card.toString()
        }
    }

}